Sources/GRPC/ConnectivityState.swift

```diff
@@ -0,0 +1,134 @@
+  /// equivalent) and protocol-level (HTTP/2, etc) handshaking.
+  case ready
+
+  /// There has been some transient failure (such as a TCP 3-way handshake timing out or a socket
+  /// error). Channels in this state will eventually switch to the `.connecting` state and try to
+  /// establish a connection again. Since retries are done with exponential backoff, channels that
+  /// fail to connect will start out spending very little time in this state but as the attempts
+  /// fail repeatedly, the channel will spend increasingly large amounts of time in this state.
+  case transientFailure
+
+  /// This channel has started shutting down. Any new RPCs should fail immediately. Pending RPCs
+  /// may continue running till the application cancels them. Channels may enter this state either
+  /// because the application explicitly requested a shutdown or if a non-recoverable error has
+  /// happened during attempts to connect communicate. Channels that enter this state never leave
+  /// this state.
+  case shutdown
+}
+
+public protocol ConnectivityStateDelegate: class {
+  /// Called when a change in `ConnectivityState` has occurred.
+  ///
+  /// - Parameter oldState: The old connectivity state.
+  /// - Parameter newState: The new connectivity state.
+  func connectivityStateDidChange(from oldState: ConnectivityState, to newState: ConnectivityState)
+}
+
+public class ConnectivityStateMonitor {
+  public typealias Callback = () -> Void
+
+  private var idleCallback: Callback?
+  private var connectingCallback: Callback?
+  private var readyCallback: Callback?
+  private var transientFailureCallback: Callback?
+  private var shutdownCallback: Callback?
+
+  /// A delegate to call when the connectivity state changes.
+  public var delegate: ConnectivityStateDelegate?
+
+  /// The current state of connectivity.
+  public internal(set) var state: ConnectivityState {
+    didSet {
+      if oldValue != self.state {
+        self.delegate?.connectivityStateDidChange(from: oldValue, to: self.state)
+        self.triggerAndResetCallback()
+      }
+    }
+  }
+
+  /// Creates a new connectivity state monitor.
+  ///
+  /// - Parameter delegate: A delegate to call when the connectivity state changes.
+  /// - Parameter initialState: The initial connectivity state, defaults to `.idle`.
+  public init(delegate: ConnectivityStateDelegate?, initialState: ConnectivityState = .idle) {
+    self.delegate = delegate
+    self.state = .idle
+  }
+
+  /// Registers a callback on the given state and calls it the next time that state is observed.
+  /// Subsequent transitions to that state will **not** trigger the callback.
+  ///
+  /// - Parameter state: The state on which to call the given callback.
+  /// - Parameter callback: The closure to call once the given state has been transitioned to. The
+  ///     `callback` can be removed by passing in `nil`.
+  public func onNext(state: ConnectivityState, callback: Callback?) {
```

Comment:
What is this for as opposed to just providing a delegate, and why is the callback reset after every change to that state?

Reply:
Good question; my take on it is that they're useful for different periods of time. You can achieve exactly the same with a delegate but in some cases this is just _much_ more convenient (the tests are a good example of this). The delegate is useful for longer time periods (logging/monitoring). This is useful for single events (e.g. my connection has shutdown, I can now update my UI) which is why it resets after each change. Definitely open to not resetting after each change though as I realise that's a slightly strange decision!

Reply:
Also, if we don't reset, we could have the user pass in a closure instead that could `switch` over the states they are interested in and discard the rest via `default:`.

At that point, one could also argue over removing the delegate and only letting the user provide callback closures. And at that point we could even keep an array of closures, in case the user wants more than one observer.

To be honest, having two different ways of passing in callback closures feels a bit weird right now, but this is not a blocker. We can revisit it later (before 1.0).